sda-commons-server-dropwizard/src/main/java/org/sdase/commons/server/dropwizard/metadata/MetadataContext.java

```diff
@@ -0,0 +1,119 @@
+package org.sdase.commons.server.dropwizard.metadata;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+/**
+ * The {@code MetadataContext} stores information about a long-running business processes that is
+ * independent of specific APIs or business rules of a single service. A {@code MetaContext} is
+ * built from the metadata fields of input APIs like HTTP or Kafka message headers. It is added
+ * automatically to output APIs when using platform clients from the {@code JerseyClientBundle} or
+ * the {@code KafkaMessageProducer}.
+ *
+ * <p>Services that claim to support {@code MetadataContext} must take care that it is also kept and
+ * reloaded when the business process is interrupted and proceeded later. An interrupted business
+ * process in these terms is for example asynchronous processing in a new {@link Thread} or
+ * finishing the current step of the process by saving the current state to a database.
+ *
+ * <p>The metadata fields (e.g. header names) that are put into the {@code MetadataContext} must be
+ * configured by {@linkplain System#getenv(String) environment variable} or {@linkplain
+ * System#getProperty(String) system property} {@code METADATA_FIELDS} as comma separated list of
+ * header names, e.g. {@code business-process-id,tenant-id}. Header names are treated
+ * case-insensitive.
+ *
+ * <p>Technically, the {@code MetadataContext} is implemented as static {@link ThreadLocal}. When
+ * code is executed in a new {@link Thread}, the context must be transferred. Helper methods are
+ * available for {@linkplain #transferMetadataContext(Runnable) <code>Runnable</code>} and
+ * {@linkplain #transferMetadataContext(Callable) <code>Callable</code>}
+ */
+// TODO not picked up from request headers and Kafka consumers as documented yet.
+// TODO Not handled in platform client and KafkaMessageProducer as documented yet.
+public interface MetadataContext {
+
+  /**
+   * @return the immutable metadata context of the current {@link Thread}, never {@code null}.
+   */
+  static MetadataContext current() {
+    return MetadataContextUtil.current();
+  }
+
+  /**
+   * @return the current metadata context, never {@code null}. Changes in the returned instance will
+   *     not affect the {@link MetadataContext} of the current {@link Thread}.
+   */
+  static DetachedMetadataContext detachedCurrent() {
+    return DetachedMetadataContext.of(MetadataContextUtil.current());
+  }
+
+  /**
+   * Creates a new {@link MetadataContext} for the current {@link Thread}. Any existing {@link
+   * MetadataContext} in the current {@link Thread} will be replaced.
```

Comment:
Could we add warning here for the replacement behaviour? or maybe away how to suppress it and passively set it if the user needs it? 
From what I understood, depending on the specific service implementation, restoring the current context from a `DetachedMetadataContext` for e.g from a database could lead to some unexpected behaviour if in the case of the process triggered from a http request with predefined metadata, the user could then define different `createContext` strategies (replace, skip, fail...) in multiple locations in the implementation

Reply:
> Especially if the current service does not know all the metadata upfront as they could evolve over time

That should not happen. All fields that are considered must be specified by [`METADATA_FIELDS`](https://github.com/SDA-SE/sda-dropwizard-commons/pull/2115/files#diff-4c7fc6f534d55a8e5faba904dda14f3e525d90d61d1b26c735c49764b1a56a31R39-R46). Only these fields will be picked up and forwarded.

> The question is if all business services involved in a business processes should know all the metadata keys that should be transported, or is it only consumers that need to act on it?

Every service involved in the process - also the ones not interested in the metadata - must know the fields. Otherwise we have to forward all headers which could leak sensitive information or break requests. Services that don't need the metadata must not implement anything. But the deployment must define `METADATA_FIELDS`. We defined that as environment variable instead of a regular configuration, so that uninterested services don't need to add properties to their `config.yml`. However, services that interrupt the process by storing data and using it later must add the context the data, even if they are not interested in it themselves.

> in the second case a nice addition could be an explicitly set parametrised behaviour (Implemented internally in the library) for only for the intersection part(commons keys), this way keeping the rest of the metadata could assured by the library, as opposed to leaving much freedom to users to build the desired metadata from the current and the DetachedMetadataContext and use it in the createContext which does not guarantee transporting everything to upstream services which could assume that downstream are just dumb transporters for the metadata.

I'm not sure if I understand this correctly. A service can't just create any context. The metadata context is something global for all services involved. So it is highly affected by configuration. The idea is like this example:

A service that is sending emails claims, that it supports to choose a template based on the metadata context.

When the feature should be used in a specific environment/cluster based on the _tenant_, all services must be configured with `METADATA_FIELDS=tenant-id,optional-other-fields` and the email service is additionally configured with `METADATA_FIELD_TEMPLATE_HINT=`tenant-id`. `METADATA_FIELD_TEMPLATE_HINT` should be part of its deployment documentation!

When choosing the template, the email service get's the hint via `MetadataContext.current().valuesByKeyFromEnvironment("METADATA_FIELD_TEMPLATE_HINT")` and will receive the `tenant-id` from the context.

The same applies for a producer. It can't just put anything to the context but must have in its own business context a property that is mapped by configuration to a specific metadata field. In some cases you may also have no explicit service setting the meta data. A tenant or otherwise specific use of a service could just get the header from an Ingress proxy based on the Domain where the service is exposed. In such case, no developer has to set fields of the context in their service.

Reply:
Okay, I got you, and it is logical and can be already seen in the implementation and the concern should be also fixed by the proposed `merge`. 
Just that if the process is interrupted (metadata saved to a database when `METADATA_FIELDS=key-1,key-2`) then at a future state, if a new configuration is introduced that adds a new `key-3` which could be not so relevant to the current service itself but to the next one, configuring all services with `METADATA_FIELDS=key-1,key-2,key-3` should still get all new values and `key-3` that the new service will start to receive should not be lost when restoring the context.
But I guess services interacting with the context should also be aware of such case and should be able to handle this themselves with the helpers, even if the behaviour is not imposed by the library.

Reply:
I hope these additions will help:

- [`isEffectivelyEmpty()`](https://github.com/SDA-SE/sda-dropwizard-commons/pull/2115/commits/cceef85599a8ccb13c638766c8d5e38300d246fe)
- [`mergeContext(newContextData, mergeStrategy)`](https://github.com/SDA-SE/sda-dropwizard-commons/pull/2115/commits/5bcbafc6d3cc224e2488122036cd900cac67d458)